dev/TeachingTip/TeachingTip.cpp

```diff
@@ -1208,249 +1255,391 @@ void TeachingTip::StartContractToClose()
             {
-                return winrt::TeachingTipPlacementMode::BottomEdgeAlignedLeft;
+                return std::tuple<winrt::TeachingTipPlacementMode, bool>(winrt::TeachingTipPlacementMode::BottomEdgeAlignedRight, false);
             }
-            else if (rightTopAvailable)
+            if (mode == winrt::TeachingTipPlacementMode::LeftEdgeAlignedTop && leftTopAvailable)
             {
-                return winrt::TeachingTipPlacementMode::RightEdgeAlignedTop;
+                return std::tuple<winrt::TeachingTipPlacementMode, bool>(winrt::TeachingTipPlacementMode::LeftEdgeAlignedTop, false);
             }
-            else if (rightBottomAvailable)
+            if (mode == winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom && leftBottomAvailable)
             {
-                return winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom;
+                return std::tuple<winrt::TeachingTipPlacementMode, bool>(winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom, false);
             }
-            else if (leftTopAvailable)
+            if (mode == winrt::TeachingTipPlacementMode::RightEdgeAlignedTop && rightTopAvailable)
             {
-                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedTop;
+                return std::tuple<winrt::TeachingTipPlacementMode, bool>(winrt::TeachingTipPlacementMode::RightEdgeAlignedTop, false);
             }
-            else if (leftBottomAvailable)
+            if (mode == winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom && rightBottomAvailable)
             {
-                return winrt::TeachingTipPlacementMode::LeftEdgeAlignedBottom;
+                return std::tuple<winrt::TeachingTipPlacementMode, bool>(winrt::TeachingTipPlacementMode::RightEdgeAlignedBottom, false);
             }
         }
     }
-    // The teaching tip wont fit anywhere, just return top.
-    return winrt::TeachingTipPlacementMode::Top;
+    // The teaching tip wont fit anywhere, set tipDoesNotFit to indicate that we should not open.
+    return std::tuple<winrt::TeachingTipPlacementMode, bool>(winrt::TeachingTipPlacementMode::Top, true);
+}
+
+std::tuple<winrt::Rect, winrt::Thickness, winrt::Thickness> TeachingTip::DetermineSpaceAroundTarget()
+{
+    auto shouldConstrainToRootBounds = ShouldConstrainToRootBounds();
+    // We use the current window/XamlRoot bounds to determine how much space is between the bottom and left edge of the target
+    // and the bottom and left edge of the window/XamlRoot.
+    auto windowBounds = m_testWindowBounds;
+    if (!m_useTestWindowBounds)
+    {
+        windowBounds = winrt::Window::Current().CoreWindow().Bounds();
+    }
+
+    // We use the screen's size to determine how much space is available between the bottom and right edges of the window/root and the screen
+    // This is only necessary if we are displaying in a popup which is not constrained to the root bounds.
+    auto screenBounds = m_testScreenBounds;
+
```

Comment:
I thought based on the discussion we weren't doing anything screen-bounds-related anymore?

Reply:
We are not, discussed in person.

---
In reply to: [269289847](https://github.com/Microsoft/microsoft-ui-xaml/pull/480#discussion_r269289847) [](ancestors = 269289847)